File: sorting/timsort.py
MINIMUM = 32


def find_minrun(n):
    r = 0
    while n >= MINIMUM:
        r |= n & 1
        n >>= 1
    return n + r


def insertion_sort(array, left, right):
    for i in range(left + 1, right + 1):
        element = array[i]
        j = i - 1
        while element < array[j] and j >= left:
            array[j + 1] = array[j]
            j -= 1
        array[j + 1] = element
    return array


def merge(array, l, m, r):
    array_length1 = m - l + 1
    array_length2 = r - m
    left = []
    right = []
    for i in range(0, array_length1):
        left.append(array[l + i])
    for i in range(0, array_length2):
        right.append(array[m + 1 + i])

    i = 0
    j = 0
    k = l

    while j < array_length2 and i < array_length1:
        if left[i] <= right[j]:
            array[k] = left[i]
            i += 1

        else:
            array[k] = right[j]
            j += 1

        k += 1

    while i < array_length1:
        array[k] = left[i]
        k += 1
        i += 1

    while j < array_length2:
        array[k] = right[j]
        k += 1
        j += 1


def timsort(array):
    n = len(array)
    minrun = find_minrun(n)

    for start in range(0, n, minrun):
        end = min(start + minrun - 1, n - 1)
        insertion_sort(array, start, end)

    size = minrun
    while size < n:

        for left in range(0, n, 2 * size):
            mid = min(n - 1, left + size - 1)
            right = min((left + 2 * size - 1), (n - 1))
            merge(array, left, mid, right)
            print(left, mid, right)
        size = 2 * size


if __name__ == '__main__':
    import random
    import time
    import tracemalloc

    t, m = [], []

    for i in range(64):
        rand = [random.randint(0, 100) for _ in range(100)]
        arr = rand.copy()
        start = time.time()
        timsort(arr)
        end = time.time()
        t.append(end - start)
        print('Time Routine:', i, sorted(rand) == arr)

    for i in range(64):
        rand = [random.randint(0, 100000) for _ in range(100000)]
        arr = rand.copy()
        tracemalloc.start()
        timsort(arr)
        current, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        m.append(peak)
        print('Memory Routine:', i, sorted(rand) == arr)

    print('[Timsort]')

    print('Best time(ms): ', min(t) * 1000)
    print('Worst time(ms): ', max(t) * 1000)
    print('Average time(ms): ', sum(t) / len(t) * 1000)

    print('Best space(bytes): ', min(m))
    print('Worst space(bytes): ', max(m))
    print('Average space(bytes): ', sum(m) / len(m))

File: cpp/AdaptivePartitionSort_Cpp/TimSort.cpp
#include "TimSort.h"
#include <iostream>
using namespace std;
#define min(x, y) (x < y ? x : y)
#define MINIMUM 32

inline int TimSort::FindMinRun(int n) {
	register int r = 0;
	while (n >= MINIMUM) {
		r |= n & 1;
		n >>= 1;
	}
	return n + r;
}

inline void TimSort::InsertionSort(int* arr, int low, int high) {
	for (register int i = low + 1; i <= high; ++i) {
		register int e = arr[i];
		register int j = i;
		while (e < arr[--j] && j >= low) arr[j + 1] = arr[j];
		arr[j + 1] = e;
	}
}

inline void TimSort::Merge(int* arr, int low, int mid, int high) {
	register int left = low;
	register int right = mid + 1;
	register int k = -1;
	register int* tmp = new int[high - low + 1];
	while (left <= mid and right <= high) {
		if (arr[left] <= arr[right]) tmp[++k] = arr[left++];
		else tmp[++k] = arr[right++];
	}

	register int i = left - 1;
	while (i < mid) tmp[++k] = arr[++i];
	i = right - 1;
	while (i < high) tmp[++k] = arr[++i];
	i = high + 1;
	k = high - low + 1;
	while (k) arr[--i] = tmp[--k];
	delete[] tmp;
}

void TimSort::Sort(int* arr, unsigned int size) {
	int minrun = FindMinRun(size);
	for (register int start = 0; start < size; start += minrun) {
		InsertionSort(arr, start, min(start + minrun - 1, size - 1));
	}
	int n = minrun;
	--size;
	while (size + 1 > n) {
		register int unit = n << 1;
		for (register int left = 0; left < size; left += unit) {
			register int mid = min(size, left + n - 1);
			register int right = min((left + 2 * n - 1), size);
			Merge(arr, left, mid, right);
		}
		n <<= 1;
	}
}
